[JENKINS-61840] Restyle buttons

fqueiruga opened the pull request:
See [JENKINS-61840](https://issues.jenkins-ci.org/browse/JENKINS-61840).

This PR can be tested using the following docker image:
`docker run --rm -ti -p 8080:8080 -e ID=4658 jenkins/core-pr-tester`

This PR restyles the buttons. The CSS API is described in detail in the JIRA issue.

To summarize:
- Buttons are restyled from scratch. Removed references to YahooUI CSS code that dealt with buttons in order to avoid overriding styles.
- Use the same CSS API that YUI buttons used. No breaking changes here.
- Created styles for 4 button variants: primary, secondary (default), danger and link (transparent). Added a _large_ version as well.
- Added support for buttons with leading and trailing buttons.
- Added support for icon-only buttons.

Screenshot:
![Captura de pantalla 2020-04-15 a las 13 36 26](https://user-images.githubusercontent.com/5738588/79333526-fb4b8b80-7f1e-11ea-8eb3-639a04d21949.png)

<details>
<summary>Sample jelly code to test the buttons</summary>

````
    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Anchor buttons</strong>

      <a href="#" class="yui-button primary">Primary</a>
      <a href="#" class="yui-button" >Default</a>
      <a href="#" class="yui-button danger">Danger</a>
      <a href="#" class="yui-button link-button">Link</a>
    </div>

    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Large anchor buttons (disabled)</strong>

      <a href="#" class="yui-button large-button disabled primary">Primary</a>
      <a href="#" class="yui-button large-button disabled">Default</a>
      <a href="#" class="yui-button large-button disabled danger">Danger</a>
      <a href="#" class="yui-button large-button disabled link-button">Link</a>
    </div>

    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Buttons with leading icons</strong>

      <span class="yui-button primary">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_done_24px" class="leading-icon" />
          <span>Primary</span>
        </button>
      </span>
      <span class="yui-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_cached_24px" class="leading-icon" />
          <span>Default</span>
        </button>
      </span>
      <span class="yui-button danger">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_report_problem_24px" class="leading-icon" />
          <span>Danger</span>
        </button>
      </span>
      <span class="yui-button link-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_help_24px" class="leading-icon" />
          <span>Link</span>
        </button>
      </span>
    </div>

    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Large buttons with leading and trailing icons</strong>

      <span class="yui-button large-button primary">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_done_24px" class="leading-icon" />
          <span>Primary</span>
        </button>
      </span>
      <span class="yui-button large-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_cached_24px" class="leading-icon" />
          <span>Default</span>
        </button>
      </span>
      <span class="yui-button large-button danger">
        <button>
          <span>Danger</span>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_report_problem_24px" class="trailing-icon" />
        </button>
      </span>
      <span class="yui-button large-button link-button">
        <button>
          <span>Link</span>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_open_in_new_24px" class="trailing-icon" />
        </button>
      </span>
    </div>

    <div style="margin-top: 2rem; display: flex; flex-wrap: wrap">
      <strong style="display:block; width: 100%; text-align: left; margin-bottom: .5rem">Normal and large icon buttons</strong>

      <span class="yui-button icon-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-social-symbol.svg#ic_notifications_24px" />
        </button>
      </span>
      <span class="yui-button primary icon-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_cached_24px" />
        </button>
      </span>
      <span class="yui-button danger icon-button large-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-social-symbol.svg#ic_notifications_24px" />
        </button>
      </span>
      <span class="yui-button link-button icon-button large-button">
        <button>
          <l:svgIcon href="${resURL}/images/material-icons/svg-sprite-action-symbol.svg#ic_cached_24px" />
        </button>
      </span>
    </div>
````
</details>

### Proposed changelog entries

* Buttons were restyled. Added support for large buttons, hyperlinks styled as buttons and icon-only buttons.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://jenkins.io/changelog/ -->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@timja 
@daniel-beck 
@oleg-nenashev 

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a JIRA issue should exist and be labeled as `lts-candidate`

fqueiruga committed:
Restyles the buttons

- Disables default YahooUI button styles
- Use YUI classes for backwards compatibility:
    - <span class="yui-button"><button>Button element</button>
- Styles are applied by default for the following input types:
    - <input type="[button | reset | submit]" />
- Add a large button variant
    - The variant is set adding the .large-button class
    - f:apply and f:submit jelly templates now have a isLarge property that allows toggling large buttons on
    - Job configuration has large buttons

fqueiruga committed:
Adds styles for hyperlinks as buttons

fqueiruga committed:
Adds styles for icons within buttons

fqueiruga committed:
Add icon buttons and update button styles

fqueiruga committed:
Fix: removed button nested within an anchor link

- Now that the anchor tags with a .yui-button class are styled like buttons, it is unnecessary to add a button within.

timja commented on core/src/main/resources/lib/form/apply.jelly:
```diff
@@ -35,8 +35,15 @@ THE SOFTWARE.
       The text of the apply button.
     </s:attribute>
+
+    <s:attribute name="isLarge">
```
```suggestion
    <s:attribute name="large">
```

timja commented on core/src/main/resources/lib/form/apply.jelly:
```diff
@@ -35,8 +35,15 @@ THE SOFTWARE.
-  <input type="button" value="${attrs.value ?: '%Apply'}" class="apply-button applyButton" name="Apply"/><!-- applyButton is legacy -->
+  <input type="button"
+         value="${attrs.value ?: '%Apply'}"
+         class="apply-button applyButton ${attrs.isLarge ? 'large-button' : ''}"
```
```suggestion
         class="apply-button applyButton ${attrs.large ? 'large-button' : ''}"
```

timja commented on core/src/main/resources/hudson/model/Job/configure.jelly:
```diff
@@ -65,8 +65,8 @@ THE SOFTWARE.
                 <!--<input type="button" name="StructureTest" value="Test" onclick="buildFormTree(this.form)" />-->
-                <f:submit value="${%Save}"/>
-                <f:apply value="${%Apply}"/>
+                <f:submit value="${%Save}" isLarge="true"/>
```
```suggestion
                <f:submit value="${%Save}" large="true"/>
```

timja commented on core/src/main/resources/hudson/model/Job/configure.jelly:
```diff
@@ -65,8 +65,8 @@ THE SOFTWARE.
-                <f:submit value="${%Save}"/>
-                <f:apply value="${%Apply}"/>
+                <f:submit value="${%Save}" isLarge="true"/>
+                <f:apply value="${%Apply}" isLarge="true" />
```
```suggestion
                <f:apply value="${%Apply}" large="true" />
```

timja commented on core/src/main/resources/lib/form/submit.jelly:
```diff
@@ -36,6 +36,14 @@ THE SOFTWARE.
       The text of the submit button. Something like "submit", "OK", etc.
     </s:attribute>
+
+    <s:attribute name="isLarge">
```
```suggestion
    <s:attribute name="large">
```

timja commented on core/src/main/resources/lib/form/submit.jelly:
```diff
@@ -36,6 +36,14 @@ THE SOFTWARE.
+  <input type="submit"
+         name="${attrs.name ?: 'Submit'}"
+         value="${attrs.value ?: '%Submit'}"
+         class="submit-button primary ${attrs.isLarge ? 'large-button' : ''}" />
```
```suggestion
         class="submit-button primary ${attrs.large ? 'large-button' : ''}" />
```

timja reviewed (commented):
From an initial look I would prefer not to call the flag isLarge,

I'll pull this locally and take a look as well

fqueiruga committed:
Rename jelly attribute isLarge to large

Co-Authored-By: Tim Jacomb <t.jacomb@kainos.com>

timja reviewed (approved):
the email notification section in /configure looks a bit funny but all else looks good

![image](https://user-images.githubusercontent.com/21194782/79338908-f5a27580-7f1f-11ea-8ff5-322775cccc48.png)

fqueiruga committed:
Fix alignment of the notepad icon on the advanced.jelly

fqueiruga commented:
I have fixed the icon alignment on the email notification section:
<img width="666" alt="Captura de pantalla 2020-04-15 a las 15 53 14" src="https://user-images.githubusercontent.com/5738588/79346093-386d4900-7f32-11ea-96fb-c03a0401f501.png">

timja commented:
nice thanks Felix

alecharp commented:
I know this might be out of scope, but we are still using `yui` prefix, when those classes, IIRC, came from the Yahoo UI library. As we are not using that anymore, maybe we could rename the classes. WDYT?

fqueiruga commented:
We could not just rename the classes because the current CSS API is being used across the whole ecosystem. There are also many cases were we will never be able to skip the `.yui-*` classes, such as menu buttons, dropdowns, etc.

Because we need to support those features and the existing code, I avoided coming up with a new interface that would never truly replace the old one, and focused on making sure it worked with existing code. On the new addition to the button catalog, hyperlink buttons, I just kept the `.yui-button` class for consistency.

That said, I'm not 100% happy with needing to wrap buttons within a span tag, so I'm considering adding an alternative. Do you have any concrete vision for what a new CSS API would be @alecharp ?

romenrg reviewed (commented):
It's great to see more things on the UI being refreshed. Thanks @fqueiruga !

This buttons change seemed a bit subtle to me at first, but after playing a bit with it now, I start to notice the improvement more and more.

One thing I noted, though, is that the "OK" button that appears on the "Create new item" page, does not seem to be updated. So far is the only one I've detected that seems to still show with the old styles.

**Screenshots from my local execution on that page:**

The "Create new item" page and the OK button:
<img width="1436" alt="Create New Item" src="https://user-images.githubusercontent.com/1831480/79364996-f815c700-7f41-11ea-9a00-3c4a079e2a19.png">


And the over effect in that button:
<img width="128" alt="Create New Item hover" src="https://user-images.githubusercontent.com/1831480/79364617-83db2380-7f41-11ea-9bc1-2ec3f9d267ca.png">

As I said, in the rest of screens I tested, I saw the new styles.

StefanSpieker reviewed (approved):
I really like the new buttons. Thanks for the restyling 👍

alecharp reviewed (approved):
I tested locally and I didn't see any problem. 

@fqueiruga for the renaming, it was just an idea. Normally, plugin devs shouldn't be calling the classes directly, but be using the jelly tags. But you are correct, we never know.

fqueiruga commented:
I think buttons, hyperlinks, etc  are UI primitives that I would expect everybody could use without repercusions. Not everything can go through jelly if that makes sense.

timja commented:
> I think buttons, hyperlinks, etc are UI primitives that I would expect everybody could use without repercusions. Not everything can go through jelly if that makes sense.

I wouldn't expect many buttons not to go through jelly, links sure...

fqueiruga committed:
Fix broken test due to changes on check.jelly

- e66df30681cddf6d6511e89a4a2c8d63833d8d67 broke this test when removing the button nested within the anchor

fqueiruga committed:
Use normal button styles on the New Item page

fqueiruga commented:
Fixed that problem Romén, re-requesting your review

romenrg reviewed (approved):
Awesome @fqueiruga .

Just tested the new changes locally and the "new item" page does indeed show the new styles in the "OK" button now as well. Thanks!

timja commented:
Looks great thanks Felix and thanks everyone for all the reviews and trying it out

We'll merge tomorrow if no negative feedback

daniel-beck commented:
Looks great.

The only minor issue I found was that the expandable multi-line input fields now look even weirder than before, but they were really ugly before already, so no big deal.

> <img width="2152" alt="Screenshot 2020-04-18 at 00 52 25" src="https://user-images.githubusercontent.com/1831569/79621017-264a0100-8112-11ea-847e-73e52982d6ce.png">

daniel-beck commented on core/src/main/resources/lib/layout/layout.jelly:
```diff
@@ -153,9 +153,7 @@ THE SOFTWARE.
     </j:if>
 
     <link rel="stylesheet" href="${resURL}/scripts/yui/container/assets/container.css" type="text/css"/>
-    <link rel="stylesheet" href="${resURL}/scripts/yui/assets/skins/sam/skin.css" type="text/css" />
```
This probably caused https://issues.jenkins-ci.org/browse/JENKINS-62001
